modules/dnn/src/tensorflow/tf_graph_simplifier.cpp

```diff
@@ -1019,11 +1019,11 @@ void sortByExecutionOrder(tensorflow::GraphDef& net)
             {
                 int numControlEdges = 0;
                 for (int j = 0; j < numInputsInGraph; ++j)
-                    numControlEdges += node.input(j)[0] == '^';
-                numRefsToAdd[i] = numControlEdges + 1;
+                    numControlEdges += node.input(j).at(0) == '^';
+                numRefsToAdd.at(i) = numControlEdges + 1;
             }
             else
-                numRefsToAdd[i] = numInputsInGraph;
+                numRefsToAdd.at(i) = numInputsInGraph;
```

Comment:
`.at()` emits on error:

```
C++ exception with description "vector::_M_range_check: __n (which is 10) >= this->size() (which is 1)" thrown in the test body.
```

OpenCV check (before `if (node.op() == "Merge" ...`)
```
CV_CheckLT((size_t)i, numRefsToAdd.size(), "");
```

could emit message with line number from OpenCV codebase.

```
C++ exception with description "OpenCV(3.4.17-dev) modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:1018: error: (-2:Unspecified error) in function 'void cv::dnn::experimental_dnn_34_v24::sortByExecutionOrder(opencv_tensorflow::GraphDef&)'
>  (expected: '(size_t)i < numRefsToAdd.size()'), where
>     '(size_t)i' is 10
> must be less than
>     'numRefsToAdd.size()' is 1
" thrown in the test body.
```

---

If user comes with report with the first message only then it would be bad for problem investigation (we need to start from beginning and ask for debugger and its call trace to confirm the problem).

`.at()` is harmless in importer code (until we have problems with importer performance), but IMHO, it is for lazy developers who don't add code for validation of input.

---

BTW, `ENABLE_GNU_STL_DEBUG` / `_GLIBCXX_DEBUG` (which enables debug checks including `[i]` case) causes crashes inside of protobuf code. Perhaps something is broken (e.g. modifying of protobuf nodes)

---

Accurate check looks like
```CV_CheckEQ(nodesMap.size(), (size_t)net.node_size(), "node names must be unique");```
after the first for loop (line 992).